src/editor/active-editor-hooks.tsx

```diff
@@ -14,28 +14,40 @@ import React, {
   useContext,
   useState,
 } from "react";
+import { ActionFeedback } from "../common/use-action-feedback";
 import { Logging } from "../logging/logging";
 import { CodeInsertType } from "./codemirror/dnd";
-import { calculateChanges } from "./codemirror/edits";
 
 /**
  * Actions that operate on a CM editor.
  */
 export class EditorActions {
-  constructor(private view: EditorView, private logging: Logging) {}
+  constructor(
+    private view: EditorView,
+    private logging: Logging,
+    private actionFeedback: ActionFeedback
+  ) {}
 
-  /**
-   * A smart, import-aware code insert.
-   *
-   * @param code The code with any required imports.
-   */
-  insertCode = (code: string, type: CodeInsertType, id?: string): void => {
+  copyCode = async (
+    code: string,
+    type: CodeInsertType,
+    id?: string
+  ): Promise<void> => {
     this.logging.event({
-      type: "code-insert",
+      type: "code-copy",
       message: id,
     });
-    this.view.dispatch(calculateChanges(this.view.state, code, type));
-    this.view.focus();
+    await navigator.clipboard.write([
+      new ClipboardItem({
+        // @ts-ignore
+        "text/html": new Blob([`<code data-type="${type}">${code}</code>`], {
```

Comment:
Presumably we should be HTML escaping the code here.
Maybe easiest to just build it as as DOM and use outerHTML?

Reply:
[`5eac9df`](https://github.com/microbit-foundation/python-editor-next/pull/766/commits/5eac9df004ad23d739064f5f22be3e9e3791e90c).